File: haskell/glue/src/Glue/Lib/Math/Utility/Round.hs
module Glue.Lib.Math.Utility.Round where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Round function
-- Mirrors Haskell Glue.Lib.Math.Utility.Round.round exactly
round :: IR Eval
round = NativeFunc roundImpl

-- Round function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Round.roundImpl exactly
roundImpl :: IR Eval -> Eval (IR Eval)
roundImpl arg = case arg of
    Integer n -> pure $ Integer n
    Float n -> pure $ Integer (Prelude.round n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/utility/round.dart
import '../../../eval.dart';
import '../../../eval/exception.dart';
import '../../../ir.dart';

/// Round function (rounds to nearest integer)
/// Mirrors Haskell Glue.Lib.Math.Utility.Round.round exactly
Eval<Ir> round(List<Ir> args) {
  return switch (args) {
    [final arg] => eval(arg).flatMap((va) {
      return switch (va) {
        IrInteger(value: final n) => Eval.pure(IrInteger(n)),
        IrFloat(value: final n) => Eval.pure(IrInteger(n.round())),
        _ => throwError(wrongArgumentType(['number'])),
      };
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}
